### ISSUE TITLE
Mean value coordinates

### ISSUE THREAD

**Issue by vchizhov** [danston/gbc#1]
I tested both of your routines for mean value coordinates (both the slower and more robust one, and the faster, but less robust one) - both seem to fail on a very simple example: a triangle with different colors at the vertices. I compared those to an implementation that I made based on Floater's paper, which seems to work correctly. Also note, that you have a mistake in the implementation of the fast routine - there is no multiplication by 0.5 for the cross product in the slides.

**Comment by danston** [comment]
Hey @vchizhov Thank you for letting me know about a possible issue in my code. In fact, I was cleaning up and simplifying my original research code right before publishing it and so I had not tested it well and it might well be that there is a problem. However, let me add a few comments:

- first, can you please give me more details on the failed case, which triangle, which coordinates, which query points, which colors, and what is your indication of being failed/being correct

- what is your implementation, Floater has many papers with different implementations tuned for different cases

- if you want to compute anything on a triangle it is completely redundant and even dangerous using any barycentric coordinates except for the triangle coordinates, which are specifically designed for that; this is because generalized coordinates always converge to triangle coordinates in the triangle case, but they have much more complicated numerical computation that may cause various numerical instabilities, while triangle coordinates do not have any of these problems

- what you called a mistake is not a mistake, 0.5 is just a scale factor, which does not affect the sign or anything important in the formula and if you look at the formula derivation, you will see that it is anyway goes away when the weights w_i are normalized, that is why you do not see it on the slides, but I have added it because theoretically A_i is the area of a triangle and not of a parallelogram that is defined by the cross product of two vectors, as you may know; in the slides it is removed to reduce the number of multiplications in the formula.

**Comment by vchizhov** [comment]
Unfortunately, I have overwritten the example that I had, but if memory serves me right, the triangle had vertices (50, 50), (450, 50), (256, 450), and colors {0, 128, 255} (greyscale).

As for Floater's paper, it is: "Generalized barycentric coordinates and
applications" - namely the code at the end for "Global form of MV coordinates". The results that I get with this code are reasonable, while the ports that I made from both of your versions create discontinuities inside the triangle.

> if you want to compute anything on a triangle it is completely redundant and even dangerous using any barycentric coordinates except for the triangle coordinates

This is untrue. Mean value coordinates work just fine for triangles - you get a smoother transition, and a more constant behaviour in the interior.

> this is because generalized coordinates always converge to triangle coordinates in the triangle case

This is also untrue.

> what you called a mistake is not a mistake, 0.5 is just a scale factor, which does not affect the sign or anything important in the formula and if you look at the formula derivation

It did change the results that I got. It is possible that I made I mistake somewhere, but I went through it several times, and I also implemented Floater's algorithm mentioned above, which works just fine.

**Comment by danston** [comment]
> while the ports that I made from both of your versions create discontinuities inside the triangle.

I need to see the results and how you used my coordinates, otherwise I cannot say anything. I can only guess that the form that you used is more stable for your points because of two possible factors: multiplication by 0.5 adds one more multiplication that in general may lead to some floating point truncations for particular points near the boundary and usage of the fabs() function in my implementation that again may create some light discontinuities near the boundary, but it is used there, because otherwise the sqrt() function there is not defined while the original Floater's weights may give tiny negative values in this case. Moreover, my function is implemented to work on general concave polygons in which case I include the mvsign() function that is what missing in the Floater's implementation, too.

> This is untrue. Mean value coordinates work just fine for triangles - you get a smoother transition, and a more constant behaviour in the interior.

I would strongly disagree with this! You are judging only form the image that you see. This is mathematically incorrect. The image is produced by which software, in which precision, with what optimizations, etc. It's all added to the final result. Theoretically mean value coordinates are identical to the triangle coordinates in the triangle. It is a linear function that is C^inf smooth, you can't say that mean value produce smoother result. In practice, in triangles, mean value coordinates only numerically converge to the triangle coordinates while the latter are optimal. But yes, different implementations matter. Some implementations may be more precise than others in some particular cases (see above). If you mean the transitions between triangles in a mesh, then what you see is super implementation dependent and this transition is only C^0 when using barycentric coordinates.

> This is also untrue.

Theoretically what I said is true. The whole idea of generalized barycentric coordinates is to **generalize** triangle coordinates to a general polygon.

Anyway, I do not see any errors in my code. And probably, your code does not contain them, too. But thanks for letting me know about these numerical imprecisions. If I have time, I will see if what I compute may be better optimized for triangles.

**Comment by vchizhov** [comment]
I'll try to come up with a benchmark one of these days to compare the different methods then. As far as the boundaries go, I do not consider any artifacts that may occur at them.

I sat down and derived the math after your comment, and indeed the coordinates should match as you noted. Which is even more bizarre considering the results I get.

I'll look into this whole thing, and I'll let you know whenever I get some results that may hint at where the problem lies.

Edit: Note that the fast algorithm seems to have been updated: https://www.inf.usi.ch/hormann/barycentric/Hormann-2011.pdf
I haven't tested the new implementation, but it probably rectifies the problems that I encountered.
